solutions/gold/cses-2421.mdx

```diff
@@ -0,0 +1,1403 @@
+	static unsigned int umod() { return bt.umod(); }
+};
+template <int id> internal::barrett dynamic_modint<id>::bt(998244353);
+
+using modint998244353 = static_modint<998244353>;
+using modint1000000007 = static_modint<1000000007>;
+using modint = dynamic_modint<-1>;
+
+namespace internal {
+
+template <class T>
+using is_static_modint = std::is_base_of<internal::static_modint_base, T>;
+
+template <class T>
+using is_static_modint_t = std::enable_if_t<is_static_modint<T>::value>;
+
+template <class> struct is_dynamic_modint : public std::false_type {};
+template <int id>
+struct is_dynamic_modint<dynamic_modint<id>> : public std::true_type {};
+
+template <class T>
+using is_dynamic_modint_t = std::enable_if_t<is_dynamic_modint<T>::value>;
+
+}  // namespace internal
+
+}  // namespace atcoder
+// EndCodeSnip
+
+const int MOD = 1'000'000'007;
+using mint = atcoder::modint1000000007;
+const int N = 26;
+
+int main() {
+	std::string s;
+	std::cin >> s;
+	int n = (int)s.size();
+	std::array<int, N> a;
+	a.fill(0);
+	for (char c : s) a[c - 'a']++;
+	int m = 0;
+	for (int i = 0; i < N; i++)
+		if (a[i] > 0) m++;
+	std::vector<mint> pp(n + 1), qq(n + 1), rr(n + 1);
+	pp[0] = qq[0] = 1;
+	for (int i = 1; i <= n; i++) {
+		pp[i] = pp[i - 1] * i;
+		qq[i] = 1 / pp[i];
+		rr[i] = qq[i - 1] * qq[i];
+	}
+
+	std::vector<mint> f(1, 0);
+	f[0] = 1;
+	for (int c = 0; c < N; c++) {
+		if (a[c] == 0) continue;
+		int ff = f.size();
+		int nn = ff + a[c];
+		std::vector<mint> g(nn, 0);
+		for (int i = 0; i < nn; i++)
+			for (int j = std::max(1, i - ff + 1); j <= std::min(i, a[c]); j++)
+				g[i] += f[i - j] * pp[a[c] - 1] * rr[j] * qq[a[c] - j];
+		f = std::move(g);
+	}
+
+	mint ans = 0;
+	f.resize(n + 1);
+	for (int k = m; k <= n; k++)
+		ans += pp[k] * f[k] * ((n - k) % 2 == 1 ? -1 : 1);
+	std::cout << ans << '\n';
+}
+```
+</CPPSection>
+</LanguageSection>
+
+**Time Complexity**: $\mathcal{O}(n^2)$
+
+
+## Solution 2 - FFT
+
+Knapsack problems, in general, can be converted to polynomial multiplication. Define $p_c$ as
+$$
+p_c(x)=\sum_{d_c=1}^{a_c}\frac{{a_c-1\choose a_c-d_c}}{d_c!}\cdot x^{d_c}
+$$
+and let $p(x)=\prod_c p_c(x)$. Then, we have that $f_{m,k}$ is equal to the coefficient of $x^k$ in $p(x)$. Each polynomial multiplication can be done efficiently using FFT. We can multiply the polynomials using a divide and conquer (as opposed to sequentially) to ensure that whenever a polynomial multiplication is performed, the two terms are roughly the same size.
+
+<Info title="Alternate Derivation">
+It's possible to derive this solution directly using generating functions, as described in [here](https://math.stackexchange.com/questions/4572019/arrangements-of-a-a-b-b-b-c-c-c-c-in-which-no-two-consecutive-letters-are-the) and [here](https://math.stackexchange.com/questions/129451/find-the-number-of-arrangements-of-k-mbox-1s-k-mbox-2s-cdots/129802#129802).
+</Info>
+
+### Implementation
+
+**Time Complexity**: $\mathcal{O}(n\log n)$
```

Comment:
it's log^2 right

Reply:
The lengths of the polynomials add up to $n$, and you only do it up to 26 times.